### PR TITLE
Metadata stamp

### LINKED ISSUE #611: Source code verification
As the ways to compile contracts get more and more complex, it becomes paramount to provide a simple way to verify the source code of a contract. Also as swarm is now more or less within reach, a way to store the source code of a contract becomes viable.

The following way to perform source code verification is hopefully forward-compatible and does not rely on timing problems or the need for a registration contract.
- At compile time, the compiler creates a json object containing metadata about the contract including the source code used to compile it, the (path-qualified) name of the contract in the source, the compiler version and for usability purposes also the ABI and documentation. This json object can be amended in the future. Its byte representation is called "the metadata".
- The hash of the metadata is stored as part of the contract creation bytecode (as `push32 metadata, pop` at the beginning of the contract). It is important to store this with the creation bytecode because the constructor is not covered if it is only stored in the deployed bytecode. Also not storing it with the deployed bytecode saves costs.
- If a user wants to verify the source code of a contract, they have to ask the ethereum node for the data that was used for creating that contract (including constructor parameters, which is very important). From that data, the hash is extracted and swarm or some other storage solution is asked for the content corresponding to that hash (i.e. the metadata). From the metadata, the complier version is extracted (the "version" might also just be a swarm-hash of the binary itself), verified that it is an "official" compiler version and the compiler (or corresponding verifier) is downloaded from swarm. Both solc and solc-js provide a verification mode: You supply the creation data and the metadata and the compiler verifies that they match and optionally outputs the constructor arguments in parsed form.
- in order to make it easier to discover the creation data, the constructor will issue a log as the first operation until there is protocol-level support for that. There is of course the option to switch off both the creation of the log and the inclusion of the metadata hash.

Problems to solve: Unlinked objects cannot provide the metadata hash because the addresess of the libraries have to be part of the metadata hash. This might not be a problem in the end.

### PR DESCRIPTION
**Pull request by chriseth**
This PR adds a swarm hash of the compilation metadata (which also includes swarm hashes for the source files) to the runtime bytecode of each contract.

Fixes #611.

### CODE REVIEW

File: libsolidity/interface/CompilerStack.cpp [hunk 1]
```diff
@@ -626,6 +650,43 @@ CompilerStack::Source const& CompilerStack::source(string const& _sourceName) co
...
+	for (auto const& r: m_remappings)
+		remappings.insert(r.context + ":" + r.prefix + "=" + r.target);
+	for (auto const& r: remappings)
+		meta["settings"]["remappings"].append(r);
+
+	for (auto const& library: m_libraries)
+		meta["settings"]["libraries"][library.first] = "0x" + toHex(library.second.asBytes());
+
+	meta["output"]["abi"] = metadata(_contract, DocumentationType::ABIInterface);
+	meta["output"]["natspec"] = metadata(_contract, DocumentationType::NatspecUser);
```
Comment (axic) [inline libsolidity/interface/CompilerStack.cpp]: I would include both natspec documents. Why do we need two documents though and not a single one?

File: libsolidity/interface/CompilerStack.cpp [hunk 2]
```diff
@@ -626,6 +650,43 @@ CompilerStack::Source const& CompilerStack::source(string const& _sourceName) co
...
+
+	Json::FastWriter writer;
+	writer.omitEndingLineFeed();
+	return writer.write(meta);
```
Comment (axic) [inline libsolidity/interface/CompilerStack.cpp]: This could use the new helper, `dev::jsonCompactPrint`.

File: libsolidity/interface/CompilerStack.cpp [hunk 3]
```diff
@@ -626,6 +650,43 @@ CompilerStack::Source const& CompilerStack::source(string const& _sourceName) co
...
+	meta["compiler"]["version"] = VersionString;
+	meta["compilationTarget"] =
+		_contract.contract->sourceUnitName() +
+		":" +
+		_contract.contract->annotation().canonicalName;
+
+	for (auto const& s: m_sources)
+	{
+		solAssert(s.second.scanner, "Scanner not available");
+		meta["sources"][s.first]["swarm"] =
```
Comment (axic) [inline libsolidity/interface/CompilerStack.cpp]: I would definitely also include a raw `keccak256` of the input.

File: libsolidity/interface/CompilerStack.cpp [hunk 4]
```diff
@@ -571,26 +595,26 @@ string CompilerStack::absolutePath(string const& _path, string const& _reference
...
 	for (auto const* dependency: _contract.annotation().contractDependencies)
-		compileContract(_optimize, _runs, *dependency, _compiledContracts);
+		compileContract(*dependency, _compiledContracts);
 
-	shared_ptr<Compiler> compiler = make_shared<Compiler>(_optimize, _runs);
-	compiler->compileContract(_contract, _compiledContracts);
+	shared_ptr<Compiler> compiler = make_shared<Compiler>(m_optimize, m_optimizeRuns);
 	Contract& compiledContract = m_contracts.at(_contract.name());
+	string onChainMetadata = createOnChainMetadata(compiledContract);
+	compiler->compileContract(_contract, _compiledContracts, dev::swarmHash(onChainMetadata));
```
Comment (axic) [inline libsolidity/interface/CompilerStack.cpp]: I would definitely include `keccak256` _and_ the swarm hash.

**Comment by axic** [comment]
I think the settings part is so much Solidity specific that either we should move them under a `solidity` node or remove the `language` node.

File: libsolidity/interface/CompilerStack.cpp (hunk 2 above)
Comment (chriseth) [reply libsolidity/interface/CompilerStack.cpp]: Yes, I have to rebase.

File: libdevcore/SwarmHash.cpp [hunk 5]
```diff
@@ -38,32 +38,28 @@ h256 swarmHashSimple(bytesConstRef _data, size_t _size)
...
-		data = std::move(innerNodes);
-		level++;
+		ref = bytesConstRef(&innerNodes);
 	}
-	while (data.size() > 32);
-	return h256(data);
+	return swarmHashSimple(ref, _length);
+}
+
+h256 dev::swarmHash(string const& _input)
```
Comment (axic) [inline libdevcore/SwarmHash.cpp:62]: I would keep both `bytes` and `string` implementations of `swarmHash`. Also it should be a separate commit and not part of the metadata change.

File: libsolidity/interface/CompilerStack.cpp (hunk 3 above)
Comment (chriseth) [reply libsolidity/interface/CompilerStack.cpp]: We can include that in the spec, but I do not really see the full point of that, at least as long as the swarm hash is still extremely similar to the sha3 hash.

File: libsolidity/interface/CompilerStack.cpp (hunk 4 above)
Comment (chriseth) [reply libsolidity/interface/CompilerStack.cpp]: Again, what would be the use of that?

File: libdevcore/SwarmHash.cpp (hunk 5 above)
Comment (chriseth) [reply libdevcore/SwarmHash.cpp:62]: We don't have a use for bytes.

**Comment by chriseth** [comment]
Concerning the layout of the settings option: Of course it is solidity specific. The keys that are not solidity specific are `version`, `language` and some parts of `output`. The specification would say that `settings` is optional, but if it is present, its structure depends on the language. Also the `compiler` key is language-specific, but I don't see a reason to introduce an extra hierarchy level for the language-specific parts.

The `compiler` key is used to retrieve a compiler for source verification. Combined with `settings` it is possible to reproduce the compilation process.

We might want to move `compilationTarget` (which is also solidity-specific) into `settings`, but it is probably better for it to be a bit more prominent.

Are there any other items we want to include?

The idea here is that we include everything that is required for recompiling and also some things that are useful without recompiling.

I did not include the source mappings and the AST for example, because if you want to use that, you usually also have a compiler to just re-generate them.

**Comment by chriseth** [comment]
Updated.

**Comment by chriseth** [comment]
Agreed to use `userdoc` and `devdoc` instead of `natspec`

**Comment by chriseth** [comment]
`compilationTarget` should move into settings

File: libsolidity/interface/CompilerStack.cpp (hunk 4 above)
Comment (axic) [reply libsolidity/interface/CompilerStack.cpp]: If I acquire the metadata through a different channel (such as a registry, or a block explorer), I can still validate it's hash without the need of understanding a specific version of Swarm's hashing algorithm (which is subject to change, whereas `keccak256` is not).

**Comment by axic** [comment]
Is https://github.com/ethereum/wiki/wiki/NatSpec-Determination superseded by this metadata proposal?

**Comment by chriseth** [comment]
Yes, the wiki page is superseded.

**Comment by chriseth** [comment]
@axic @pirapira please review, this would be my final version for now.

File: docs/miscellaneous.rst [hunk 6]
```diff
@@ -221,6 +221,125 @@ Either add ``--libraries "Math:0x12345678901234567890 Heap:0xabcdef0123456"`` to
...
+      // Required: Details about the compiler, contents are specific
+      // to the language.
+      compiler: {
+        // Required for Solidity: Version of the compiler
+        version: "0.4.6+commit.2dabbdf0.Emscripten.clang",
+        // Optional: Hash of the compiler binary which produced this output
+        keccak256: "0x123..."
+      },
+      // Required: Compilation source files/source units, keys are file names
+      sources:
```
Comment (pirapira) [inline docs/miscellaneous.rst:279]: Can the ordering of input files affect the outcome?

File: docs/miscellaneous.rst [hunk 7]
```diff
@@ -221,6 +221,125 @@ Either add ``--libraries "Math:0x12345678901234567890 Heap:0xabcdef0123456"`` to
...
+        "mortal": {
+          "keccak256": "0x123...",
+          // Required (unless "url" is used): literal contents of the source file
+          "content": "contract mortal is owned { function kill() { if (msg.sender == owner) selfdestruct(owner); } }"
+        }
+      },
+      // Required: Compiler settings
+      settings:
+      {
+        // Required for Solidity: Sorted list of remappings
```
Comment (pirapira) [inline docs/miscellaneous.rst:299]: Command line arguments in `argv` would be useful for any compiler (this can be deferred to version 2 of this format).

File: docs/miscellaneous.rst [hunk 8]
```diff
@@ -221,6 +221,125 @@ Either add ``--libraries "Math:0x12345678901234567890 Heap:0xabcdef0123456"`` to
...
+and NatSpec documentation in order to more safely interact with the contract
+and to verify its source code.
+
+The compiler inserts a swarm hash of that file into the bytecode of each
+contract, so that you can retrieve the file in an authenticated way
+without having to resort to a centralized data provider.
+
+Specifically, the runtime code for a contract always starts with
+``push32 <metadata hash> pop``, so you can take a look at the 32 bytes starting at
+the second byte of the code of a contract.
```
Comment (pirapira) [inline docs/miscellaneous.rst]: Can this be put in the last 33 bytes of the runtime code?  Then we save some runtime gas.

File: libdevcore/SwarmHash.h [hunk 9]
```diff
@@ -20,12 +20,13 @@
 #pragma once
 
 #include <libdevcore/FixedHash.h>
-#include <libdevcore/Common.h>
+
+#include <string>
```
Comment (pirapira) [inline libdevcore/SwarmHash.h:24]: nitpick: the coding rule says STL includes should be placed above `libdevcore/*`.

File: libsolidity/ast/AST.cpp [hunk 10]
```diff
@@ -63,6 +63,14 @@ SourceUnitAnnotation& SourceUnit::annotation() const
 	return static_cast<SourceUnitAnnotation&>(*m_annotation);
 }
 
+string Declaration::sourceUnitName() const
+{
+	ASTNode const* scope = m_scope;
+	while (dynamic_cast<Declaration const*>(scope) && dynamic_cast<Declaration const*>(scope)->m_scope)
+		scope = dynamic_cast<Declaration const*>(scope)->m_scope;
+	return dynamic_cast<SourceUnit const&>(*scope).annotation().path;
```
Comment (pirapira) [inline libsolidity/ast/AST.cpp:72]: `scope` here might be `nullptr` when `m_scope` is `nullptr`.

File: libsolidity/codegen/CompilerContext.cpp [hunk 11]
```diff
@@ -227,6 +227,13 @@ void CompilerContext::injectVersionStampIntoSub(size_t _subIndex)
 	sub.injectStart(fromBigEndian<u256>(binaryVersion()));
 }
 
+void CompilerContext::injectMetadataHashIntoSub(size_t _subIndex, h256 const& _metadataHash)
+{
+	eth::Assembly& sub = m_asm->sub(_subIndex);
+	sub.injectStart(Instruction::POP);
+	sub.injectStart(u256(_metadataHash));
```
Comment (pirapira) [inline libsolidity/codegen/CompilerContext.cpp]: When the number `_metadataHash` is accidentially small, this could produce `PUSH31` instead of `PUSH32`.

**Review by pirapira — CHANGES_REQUESTED** [review]

File: libsolidity/interface/CompilerStack.cpp [hunk 12]
```diff
@@ -356,32 +367,47 @@ Json::Value const& CompilerStack::metadata(string const& _contractName, Document
...
+		doc = &_contract.interface;
 		break;
 	default:
 		BOOST_THROW_EXCEPTION(InternalCompilerError() << errinfo_comment("Illegal documentation type."));
 	}
 
 	// caches the result
 	if (!*doc)
-		doc->reset(new Json::Value(InterfaceHandler::documentation(*currentContract.contract, _type)));
+		doc->reset(new Json::Value(InterfaceHandler::documentation(*_contract.contract, _type)));
```
Comment (pirapira) [inline libsolidity/interface/CompilerStack.cpp:399]: `_contract.contract` could be a `nullptr`.

File: docs/miscellaneous.rst (hunk 6 above)
Comment (chriseth) [reply docs/miscellaneous.rst:279]: No, they are sorted by name in the compiler and also here, but good point!

File: docs/miscellaneous.rst (hunk 7 above)
Comment (chriseth) [reply docs/miscellaneous.rst:299]: We are trying to go away from restricting us to a commandline interface for compiler input. Instead, we want to go to a json object for the settings, which would be compatible with this one.

File: docs/miscellaneous.rst (hunk 8 above)
Comment (chriseth) [reply docs/miscellaneous.rst]: I have thought about that, too. Currently, this costs 5 gas, which is minimal but might add up to quite a bit.
I will take a look and check if that is doable.

File: libdevcore/SwarmHash.h (hunk 9 above)
Comment (chriseth) [reply libdevcore/SwarmHash.h:24]: Let's change the coding rule: http://stackoverflow.com/a/614333/5197297 (read the full answer)

File: libsolidity/codegen/CompilerContext.cpp (hunk 11 above)
Comment (chriseth) [reply libsolidity/codegen/CompilerContext.cpp]: Good find! It might be quite hard to force that to `push32`. Would it be better to change the documentation/specification?
On the other hand, if we add that to the end of the bytecode, then there might be another way to do it.

File: libdevcore/SwarmHash.h (hunk 9 above)
Comment (pirapira) [reply libdevcore/SwarmHash.h:24]: OK.  I support the change.

File: libsolidity/codegen/CompilerContext.cpp (hunk 11 above)
Comment (pirapira) [reply libsolidity/codegen/CompilerContext.cpp]: Either way is fine: changing the documentation or forcing `PUSH32`.

**Comment by chriseth** [comment]
Metadata is now appended to the end of the runtime code. All other comments are also addressed.

File: docs/miscellaneous.rst (hunk 6 above)
Comment (pirapira) [reply docs/miscellaneous.rst:279]: OK.  Other compilers can add their own fields indicating the order of sources.

File: docs/miscellaneous.rst (hunk 7 above)
Comment (pirapira) [reply docs/miscellaneous.rst:299]: I see.

File: libsolidity/codegen/CompilerContext.cpp (hunk 11 above)
Comment (pirapira) [reply libsolidity/codegen/CompilerContext.cpp]: I saw the change.  It looks good.

**Review by pirapira — APPROVED** [review]

**Comment by axic** [comment]
My comments on the other PR relating to this are still not addressed 😉

**Comment by chriseth** [comment]
@axic sorry, I could not find your comments. Can you please repeat them here?

**Comment by chriseth** [comment]
Proposal: add magic number that specifies the encoding of the metadata hash. Options include `bzzr0` - should ask swarm people about how they want to version their hash. We could also cbor-encode the magic number and the hash.

**Comment by chriseth** [comment]
@axic CBOR encoding of `["bzzr","abtttthhhhhhhhhhhhhhtttttttttttc"]` and `{"bzzr","abtttthhhhhhhhhhhhhhtttttttttttc"}` both need 40 bytes, so 4 bytes more than the actual payload. The problem with cbor is also that we need to find the start of the cbor encoding - not sure if that is doable at all. So I think we have to find our own modified encoding anyway.

If we want to stay flexible, I see some options:

 - reverse CBOR
 - last byte is length (where values larger than 127 are reselved), up to that, cbor encoded
 - last byte is length (...), up to that magic-number prefix plus data

**Comment by axic** [comment]
> The problem with cbor is also that we need to find the start of the cbor encoding

That is a valid concern and I would agree the easiest is to leave a trailing fixed-length size field. I would argue to make it 16 bit wide.

Regarding the actual CBOR content I was thinking about an array, starting with a flag byte (integer):
- bit 0 could mean "Swarm hash of metadata"
- bit 1 could mean "keccak256 hash of metadata"

**Comment by axic** [comment]
As discussed, a good solution currently seems to:
- place CBOR encoded data immediately following the bytecode
- use a CBOR hashmap
- the `bzzr0` key has the current Swarm hash as the value
- the CBOR is followed by a 16 bit big endian length field, representing the total size of the CBOR encoded data, excluding this length field

File: docs/miscellaneous.rst [hunk 13]
```diff
@@ -319,13 +321,29 @@ Comments are of course also not permitted and used here only for explanatory pur
...
+---------------------------------------------
+
+Because we might support other ways to retrieve the metadata file in the future,
+the mapping ``{"bzzr0": <swarm hash>}`` is stored
+[CBOR](https://tools.ietf.org/html/rfc7049)-encoded. Since the beginning of that
+encoding is not easy to find, its length is added in a two-byte big-endian
+encoding. The current version of the Solidity compiler thus adds the following
+to the end of the deployed bytecode::
+
+    0xa1 0x65 'b' 'z' 'z' 'r' '0' 0x58 0x20 <32 bytes swarm hash> 0x00 0x29
```
Comment (axic) [inline docs/miscellaneous.rst]: 0x29 -> 0x1d as the length is 29 in decimal

File: libsolidity/codegen/CompilerContext.cpp [hunk 14]
```diff
@@ -220,13 +220,6 @@ void CompilerContext::appendInlineAssembly(
 	solAssert(assembly::InlineAssemblyStack().parseAndAssemble(*assembly, *m_asm, identifierAccess), "");
 }
 
-void CompilerContext::injectVersionStampIntoSub(size_t _subIndex)
-{
-	eth::Assembly& sub = m_asm->sub(_subIndex);
-	sub.injectStart(Instruction::POP);
-	sub.injectStart(fromBigEndian<u256>(binaryVersion()));
```
Comment (axic) [inline libsolidity/codegen/CompilerContext.cpp:227]: Should we include this in the CBOR?

File: test/libsolidity/SolidityABIJSON.cpp [hunk 15]
```diff
@@ -731,6 +734,23 @@ BOOST_AUTO_TEST_CASE(function_type)
...
+		contract test {
+			function g(function(uint) external returns (uint) x) {}
+		}
+	)";
+	BOOST_REQUIRE(m_compilerStack.compile(std::string(sourceCode)));
+	bytes const& bytecode = m_compilerStack.runtimeObject("test").bytecode;
+	bytes hash = dev::swarmHash(m_compilerStack.onChainMetadata("test")).asBytes();
+	BOOST_REQUIRE(hash.size() == 32);
+	BOOST_REQUIRE(bytecode.size() >= hash.size());
+	BOOST_CHECK(std::equal(hash.begin(), hash.end(), bytecode.end() - 32));
```
Comment (axic) [inline test/libsolidity/SolidityABIJSON.cpp]: This needs to be updated for CBOR.

File: libdevcore/SwarmHash.cpp [hunk 16]
```diff
@@ -38,13 +38,14 @@ h256 swarmHashSimple(bytesConstRef _data, size_t _size)
 	return keccak256(toLittleEndian(_size) + _data.toBytes());
 }
 
-h256 swarmHashIntermediate(bytes const& _input, size_t _offset, size_t _length)
+h256 swarmHashIntermediate(string const& _input, size_t _offset, size_t _length)
```
Comment (axic) [inline libdevcore/SwarmHash.cpp:41]: I would still like to keep both `bytes` and `string` options. I think it will be useful to hash other kind of data.

**Review by axic — CHANGES_REQUESTED** [review]

**Comment by chriseth** [comment]
@axic added a PR to fix the tests.

File: libsolidity/codegen/CompilerContext.cpp (hunk 14 above)
Comment (chriseth) [reply libsolidity/codegen/CompilerContext.cpp:227]: As we said during the call, let's do it once it is necessary.

File: libdevcore/SwarmHash.cpp (hunk 16 above)
Comment (chriseth) [reply libdevcore/SwarmHash.cpp:41]: Again, let's add the other option back when we need it, currently we only need this one function.

Comment (axic) [reply libdevcore/SwarmHash.cpp:41]: Wondering why is it better changing the code back and forth in that case? 😃

**Comment by axic** [comment]
We have a test failure:

> /home/travis/build/ethereum/solidity/test/libsolidity/SolidityOptimizer.cpp(75): error in "unused_expressions": Optimizer did not reduce bytecode size.
> /home/travis/build/ethereum/solidity/test/libsolidity/SolidityOptimizer.cpp(75): error in "dead_code_elimination_across_assemblies": Optimizer did not reduce bytecode size.

**Comment by axic** [comment]
Locally for some reason this isn't compiling:
```
libsolidity/interface/CompilerStack.cpp:691:42: error: conversion from 'const size_t' (aka 'const unsigned long') to 'Json::Value' is ambiguous
        meta["settings"]["optimizer"]["runs"] = m_optimizeRuns;
                                                ^~~~~~~~~~~~~~
```

**Comment by chriseth** [comment]
Result of investigation on my local machine: I can reproduce the test failure with clang, but if I compile the contract manually with optimization with clang and gcc, I get the same result.
This could mean that the bug is in the test framework.

**Comment by axic** [comment]
Since a `STOP` is emitted before the aux data, should we mention that in the hash description? e.g. the CBOR data should be preceded by `0`.

File: libsolidity/interface/CompilerStack.h [hunk 17]
```diff
@@ -113,14 +113,15 @@ class CompilerStack: boost::noncopyable
...
 	/// @returns false on error.
-	bool compile(bool _optimize = false, unsigned _runs = 200);
+	bool compile(
+		bool _optimize = false,
+		unsigned _runs = 200,
+		std::map<std::string, h160> const& _libraries = std::map<std::string, h160>{}
+	);
 	/// Parses and compiles the given source code.
 	/// @returns false on error.
 	bool compile(std::string const& _sourceCode, bool _optimize = false);
```
Comment (axic) [inline libsolidity/interface/CompilerStack.h]: Shouldn't this also have the `_runs` parameter? Is there a need two support both APIs (this and the preceding one)?

File: libsolidity/interface/CompilerStack.h [hunk 18]
```diff
@@ -252,6 +257,9 @@ class CompilerStack: boost::noncopyable
 	};
 
 	ReadFileCallback m_readFile;
+	bool m_optimize = false;
+	unsigned m_optimizeRuns = 200;
```
Comment (axic) [inline libsolidity/interface/CompilerStack.h:261]: Probably these two shouldn't have a default here as they are set in `compile`?

**Review by axic — CHANGES_REQUESTED** [review]
There should be a test for the content of the metadata JSON. The test contract should also use a library.

**Comment by chriseth** [comment]
@axic I think verifying the content of the metadata output can be left for the pull request that introduces the input format.

File: libsolidity/interface/CompilerStack.h (hunk 18 above)
Comment (chriseth) [reply libsolidity/interface/CompilerStack.h:261]: I try to always set a default value for PODs to keep determinism.

**Comment by chriseth** [comment]
Concerning the STOP: We cannot rely on the fact that there is no other data before the metadata cbor.

**Comment by chriseth** [comment]
BUILDS ARE GREEEN! *jump*
